File: 207-course-schedule/course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        adj = [ [] for i in range(numCourses) ]
        inDegree = [0 for i in range(numCourses) ]

        for term in prerequisites:
            adj[term[1]].append(term[0])
            inDegree[term[0]] += 1

        q = []
        for i in range(numCourses):
            if inDegree[i] == 0:
                q.append(i)

        
        visit = 0
        while q:
            node = q.pop()
            visit += 1

            for nbr in adj[node]:
                inDegree[nbr] -= 1
                if inDegree[nbr] == 0:
                    q.append(nbr)

        if numCourses == visit:
            return True
        return False

        

File: 207-course-schedule/course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) 
    {
        int count = 0;

        ArrayList<ArrayList<Integer>> adj = new ArrayList<>();
        for(int i=0 ; i < numCourses ;i++)
        {
            adj.add(new ArrayList<>());
        }

        int[] inDeg = new int[numCourses];
        for(int[] term : prerequisites)
        {
            int course = term[0];
            int preq = term[1];

            adj.get(preq).add(course);
            inDeg[course]++;
        }

        Queue<Integer> q = new ArrayDeque<>();

        for(int i=0 ;i < inDeg.length ;i++)
        {
            if ( inDeg[i] == 0 )
            {
                q.offer(i);
            }
        }

        while ( !q.isEmpty())
        {
            int node = q.poll();
            count++;

            for(int nbr  : adj.get(node))
            {
                inDeg[nbr]--;
                if ( inDeg[nbr] == 0)
                {
                    q.offer(nbr);
                }                
            }
        }

        return count == numCourses;


        
    }
}